kernel/nativecode.ml

```diff
@@ -52,6 +52,11 @@ let fresh_lname n =
   incr lname_ctr;
   { lname = n; luid = !lname_ctr }
 
+let is_lazy t =
+  match Constr.kind t with
+  | App _ | LetIn _ | Case _ | Proj _ -> true
+  | _ -> false
```

Comment:
Now that I look at it, this function seems especially fragile, since the default case is `_ -> false`. It should have been the opposite: it is the negative cases that should have been listed explicitly, so that new constructors safely fall in the positive case. (Either that or both positive and negative cases should have been listed, in order to force people to think about the categorization of new constructors.)

In particular, it seems like a serious bug that `Array` is not marked as positive. (And possibly `Cast` is just as bad.)

Reply:
I confirm that it is a bug that `Cast` is not there. As for `Array`, I am pretty sure that it is a bug too, but I cannot trigger it yet because `Array` actually triggers an even worse bug. (As they say, it is the gift that keeps on giving...)